typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1302 @@
+    | SharedUnique of SharedUnique.t
+
+  let to_string = function
+    | Unused -> "unused"
+    | BorrowedShared s -> BorrowedShared.to_string s
+    | SharedUnique s -> SharedUnique.to_string s
+
+  let print ppf t = Format.fprintf ppf "%s" (to_string t)
+
+  let extract_occurrence = function
+    | Unused -> assert false
+    | BorrowedShared t -> BorrowedShared.extract_occurrence t
+    | SharedUnique t -> SharedUnique.extract_occurrence t
+
+  (* parallel composition, for composing multiple control flows *)
+  let par m0 m1 =
+    match (m0, m1) with
+    | Unused, m | m, Unused -> m
+    | BorrowedShared m0, BorrowedShared m1 ->
+        BorrowedShared (BorrowedShared.par m0 m1)
+    | BorrowedShared _, m | m, BorrowedShared _ ->
+        m (* m must be sharedunique *)
+    | SharedUnique m0, SharedUnique m1 -> SharedUnique (SharedUnique.par m0 m1)
+
+  (* sequential composition *)
+  let seq m0 m1 =
+    match (m0, m1) with
+    | Unused, m | m, Unused -> m
+    | BorrowedShared m0, BorrowedShared m1 ->
+        BorrowedShared (BorrowedShared.seq m0 m1)
+    | BorrowedShared m0, SharedUnique m1 -> (
+        match (m0, m1) with
+        | Borrowed _, Shared _ -> SharedUnique m1
+        | Borrowed _, MaybeUnique _ -> SharedUnique m1
+        | MaybeShared _, Shared _ -> SharedUnique m1
+        | MaybeShared l0, MaybeUnique l1 ->
+            (* Four cases:
+               B;S = S
+               B;U = U
+               S;S = S
+               S;U /=
+
+               We are in a dilemma: recall that B->S allows code motion, and
+               S->U allows unique overwriting. We can't have both. We first note
+               is that the first is a soft optimization, and the second is a
+               hard requirement.
+
+               A reasonable solution is thus to check if the RHS actually needs
+               to use the "unique" capabilities. If not, there is no need to
+               relax it to unique, and we will make it shared, and make LHS
+               shared for code-motion. However, there is no good way to do that,
+               because the "unique_use" in "maybe_unique" is not complete,
+               because the type-checking and uniqueness analysis is performed on
+               a per-top-level-expr basis.
+
+               Our solution is to record on the l0 that it is constrained by the
+               l1. I.e. if any of l1 is U, then each of l0 cannot be S. After
+               the type checking of the whole file, l1 will correctly tells
+               whether it needs to be unique, and by extension whether l0 can be
+               shared. *)
+            let uniqs = List.map (fun ((uniq, _), _) -> uniq) l1 in
+            (* if any of l1 is unique, then all of l0 must be borrowed *)
+            let uniq = Mode.Uniqueness.meet uniqs in
+            List.iter
+              (fun (barrier, _) ->
+                match !barrier with
+                | Some _ -> assert false
```

Comment:
Yeah I believe each barrier (associate to each field projection) is written to only once. The content is strictly the "maybe unique" usage immediately following the field projection. I think it's a nice property and I should make it explicit by `assert false`. 

Also, I'm not familiar with the code-motion, but I wonder if it's flexible to express "code-motion but not beyond this point" instead of "no code-motion".

> Could we just have a list of modes on the barrier and treat them as a meet? It might be more efficient that calling actual meet anyway, especially if we included some short-circuiting for the case of constant modes.